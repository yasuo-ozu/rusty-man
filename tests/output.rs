// SPDX-FileCopyrightText: 2020 Robin Krahl <robin.krahl@ireas.org>
// SPDX-License-Identifier: MIT

use std::env;
use std::process;
use std::sync;

use assert_cmd::prelude::*;

static mut DIR: Option<tempfile::TempDir> = None;
static INIT: sync::Once = sync::Once::new();

fn run_cargo_doc() {
    INIT.call_once(|| {
        let dir = tempfile::tempdir().unwrap();
        process::Command::new(env::var_os("CARGO").unwrap())
            .arg("doc")
            .args(&["--package", "anyhow"])
            .args(&["--package", "log"])
            .args(&["--package", "rand_core"])
            .arg("--target-dir")
            .arg(dir.path())
            .output()
            .unwrap();
        unsafe {
            DIR = Some(dir);
        }
    });
}

fn run(args: &[&str]) -> assert_cmd::assert::Assert {
    run_cargo_doc();

    let dir = unsafe { DIR.as_ref().unwrap() };

    process::Command::cargo_bin(env!("CARGO_PKG_NAME"))
        .unwrap()
        .args(&["--no-default-sources", "--source"])
        .arg(dir.path().join("doc"))
        .args(&["--viewer", "plain"])
        .args(&["--width", "100"])
        .args(args)
        .assert()
}

fn get_stdout(args: &[&str]) -> String {
    let cmd = run(args).success().stderr("");
    String::from_utf8(cmd.get_output().stdout.clone()).unwrap()
}

fn assert_doc(item: &str) {
    insta::assert_snapshot!(get_stdout(&[item]));
}

fn assert_examples(item: &str) {
    insta::assert_snapshot!(get_stdout(&["-e", item]));
}

macro_rules! assert_doc {
    ($( $( #[$attrs:meta] )* $name:ident: $item:literal ),* $(,)? ) => {
        $(
            #[test]
            $( #[$attrs] )*
            fn $name() {
                assert_doc($item);
            }
        )*
    };
}

macro_rules! assert_examples {
    ($( $( #[$attrs:meta] )* $name:ident: $item:literal ),* $(,)? ) => {
        $(
            #[test]
            $( #[$attrs] )*
            fn $name() {
                assert_examples($item);
            }
        )*
    };
}

assert_doc![
    test_mod_anyhow: "anyhow",
    test_macro_anyhow_anyhow: "anyhow::anyhow",
    test_macro_anyhow_ensure: "anyhow::ensure",
    #[ignore]
    test_struct_anyhow_error: "anyhow::Error",
    test_trait_anyhow_context: "anyhow::Context",
    test_type_anyhow_result: "anyhow::Result",
];

assert_doc![
    #[ignore]
    test_mod_log: "log",
    #[ignore]
    test_macro_log_debug: "log::debug",
    #[ignore]
    test_struct_log_metadata: "log::Metadata",
    test_enum_log_level: "log::Level",
    #[ignore]
    test_constant_log_static_max_level: "log::STATIC_MAX_LEVEL",
    test_trait_log_log: "log::Log",
    test_fn_log_logger: "log::logger",
    test_fn_log_set_logger_racy: "log::set_logger_racy",
];

assert_doc![
    test_mod_rand_core: "rand_core",
    test_trait_rand_core_rngcore: "rand_core::RngCore",
    test_trait_rand_core_seedablerng: "rand_core::SeedableRng",
    test_struct_rand_core_block_blockrng: "rand_core::block::BlockRng",
];

assert_examples![
    test_examples_mod_anyhow: "anyhow",
    #[ignore]
    test_examples_mod_log: "log",
    test_examples_struct_rand_core_rngcore: "rand_core::RngCore",
];
