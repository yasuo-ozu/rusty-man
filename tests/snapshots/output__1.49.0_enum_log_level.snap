---
source: tests/output.rs
expression: "get_stdout(&[item])"
---
log                                        Enum log::Level                                 rusty-man

SYNOPSIS
      #[repr(usize)]
      pub enum Level {
          Error,
          Warn,
          Info,
          Debug,
          Trace,
      }

DESCRIPTION
      An enum representing the available verbosity levels of the logger.
      
      Typical usage includes: checking if a certain `Level` is enabled with `log_enabled!`,
      specifying the `Level` of `log!`, and comparing a `Level` directly to a `LevelFilter`.

VARIANTS
      Error
            Error

            The "error" level.
            
            Designates very serious errors.

      Warn
            Warn

            The "warn" level.
            
            Designates hazardous situations.

      Info
            Info

            The "info" level.
            
            Designates useful information.

      Debug
            Debug

            The "debug" level.
            
            Designates lower priority information.

      Trace
            Trace

            The "trace" level.
            
            Designates very low priority, often extremely verbose, information.

METHODS
   impl Level
      max
            pub fn max() -> Level

            Returns the most verbose logging level.

      to_level_filter
            pub fn to_level_filter(&self) -> LevelFilter

            Converts the `Level` to the equivalent `LevelFilter`.

IMPLEMENTATIONS
   Trait Implementations
      Clone
            impl Clone for Level

      Copy
            impl Copy for Level

      Debug
            impl Debug for Level

      Display
            impl Display for Level

      Eq
            impl Eq for Level

      FromStr
            impl FromStr for Level

      Hash
            impl Hash for Level

      Ord
            impl Ord for Level

      PartialEq
            impl PartialEq<Level> for Level

      PartialEq
            impl PartialEq<Level> for LevelFilter

      PartialEq
            impl PartialEq<LevelFilter> for Level

      PartialOrd
            impl PartialOrd<Level> for Level

      PartialOrd
            impl PartialOrd<Level> for LevelFilter

      PartialOrd
            impl PartialOrd<LevelFilter> for Level

      StructuralEq
            impl StructuralEq for Level

   Auto Trait Implementations
      RefUnwindSafe
            impl RefUnwindSafe for Level

      Send
            impl Send for Level

      Sync
            impl Sync for Level

      Unpin
            impl Unpin for Level

      UnwindSafe
            impl UnwindSafe for Level

   Blanket Implementations
      Any
            impl<T> Any for T 
            where
                T: 'static + ?Sized,

      Borrow
            impl<T> Borrow<T> for T 
            where
                T: ?Sized,

      BorrowMut
            impl<T> BorrowMut<T> for T 
            where
                T: ?Sized,

      From
            impl<T> From<T> for T

      Into
            impl<T, U> Into<U> for T 
            where
                U: From<T>,

      ToOwned
            impl<T> ToOwned for T 
            where
                T: Clone,

      ToString
            impl<T> ToString for T 
            where
                T: Display + ?Sized,

      TryFrom
            impl<T, U> TryFrom<U> for T 
            where
                U: Into<T>,

      TryInto
            impl<T, U> TryInto<U> for T 
            where
                U: TryFrom<T>,


